external/concealer/src/autoware.cpp

```diff
@@ -30,66 +33,17 @@ namespace concealer
 
 Autoware::~Autoware()
 {
-  AUTOWARE_INFO_STREAM("Shutting down Autoware: (1/3) Stop publlishing/subscribing.");
+  AUTOWARE_INFO_STREAM("Shutting down Autoware: (1/2) Stop publlishing/subscribing.");
   {
     if (spinner.joinable()) {
       promise.set_value();
       spinner.join();
     }
   }
 
-  AUTOWARE_INFO_STREAM("Shutting down Autoware: (2/3) Send SIGINT to Autoware launch process.");
-  {
-    ::kill(process_id, SIGINT);
-  }
-
-  AUTOWARE_INFO_STREAM("Shutting down Autoware: (2/3) Terminating Autoware.");
+  AUTOWARE_INFO_STREAM("Shutting down Autoware: (2/2) Send SIGINT to Autoware launch process.");
   {
-    sigset_t mask{};
-    {
-      sigset_t orig_mask{};
-
-      sigemptyset(&mask);
-      sigaddset(&mask, SIGCHLD);
-
-      if (sigprocmask(SIG_BLOCK, &mask, &orig_mask) < 0) {
-        AUTOWARE_SYSTEM_ERROR("sigprocmask");
-        std::exit(EXIT_FAILURE);
-      }
-    }
-
-    timespec timeout{};
-    {
-      timeout.tv_sec = 5;
-      timeout.tv_nsec = 0;
-    }
-
-    while (sigtimedwait(&mask, NULL, &timeout) < 0) {
-      switch (errno) {
-        case EINTR:  // Interrupted by a signal other than SIGCHLD.
-          break;
-
-        case EAGAIN:
-          AUTOWARE_ERROR_STREAM(
-            "Shutting down Autoware: (2/3) Autoware launch process does not respond. Kill it.");
-          kill(process_id, SIGKILL);
-          break;
-
-        default:
-          AUTOWARE_SYSTEM_ERROR("sigtimedwait");
-          std::exit(EXIT_FAILURE);
-      }
-    }
-  }
-
-  AUTOWARE_INFO_STREAM("Shutting down Autoware: (3/3) Waiting for Autoware to be exited.");
-  {
-    int status = 0;
-
-    if (waitpid(process_id, &status, 0) < 0) {
-      AUTOWARE_SYSTEM_ERROR("waitpid");
-      std::exit(EXIT_FAILURE);
-    }
+    sudokill(process_id);
   }
```

Comment:
The ugly code before the change played an important role.

First, if you simply kill the process that manages Autoware (`python3 ros2 launch ...`), the nodes launched by `python3 ros2 launch ...` will become zombies.

When ros2 launch receives a SIGINT, it sends a SIGINT to all the nodes it manages, and if it does not finish within 5 seconds, it transitions to a SIGKILL send.

However, if any node segfaults on termination, the ros2 launch process will never terminate because ros2 launch cannot determine if the node exited after SIGKILL.

(Sadly, SEGV at the end of Autoware happens quite often)

With the above in mind, you need to take the following steps to ensure that Autoware is shut down.

1. `python3 ros2 launch ...` Send SIGINT to the process
2. Wait for `ros2 launch` to send SIGKILL to the child process (= wait 5 seconds) and then send SIGKILL to the` ros2 launch` process itself.
3. Confirm the end of the `ros2 launch` process

Did these steps cause problems for Autoware.Auto?

Reply:
Actually the code that was in the destructor of Autoware before was not working properly for Autoware.Auto - it was only killing launch process, without proper shutdown of child processes (planning, perception etc. nodes). Thanks to replacement to sudokill method it now works properly - kills launch process, as well as all child processes. We can still refactor it in similar way that it was before if you prefer - basically sudokill method is replacement of sending SIGINT to ros2 launch.

Reply:
It's been refactor to have all the logic you implemented before, but to work with Autoware.Auto as well.

Reply:
I tried to run the refactored code with ArchitectureProposal, but sudokill still did not work properly with ArchitectureProposal.

I feel that we are facing a very complicated problem. However, since we have found code that works for both ArchitectureProposal and Auto, we think it is not worth the time to analyze this problem.

Therefore, we will postpone solving this problem.
In other words, make the destructor branch with an ifdef, and modify ArchitectureProposal to use the original code and Auto to use sudokill to do the shutdown.

Reply:
Do I understand correctly that previous simplified version with sudokill worked both for AutowareAuto and ArchitectureProposal, and new refactored version supports only AutowareAuto ?

Reply:
@yamacir-kit I addressed your comment. I branched killing Autoware for AAP (original code) and AA (`sudokill`) with `ifdef`s.

Reply:
@WJaworskiRobotec

> Do I understand correctly that previous simplified version with sudokill worked both for AutowareAuto and ArchitectureProposal, and new refactored version supports only AutowareAuto ?

No, this is actually the case.

- Original code: Works correctly with AAP, but does not work properly with AA
- sudokill (first version): Doesn't work properly with AAP, but works correctly with AA
- sudokill (refactored version): Doesn't work properly with AAP, but works correctly with AA